### PR TITLE
feat: When installing an extension, register it in the registry

### PR DESCRIPTION
**Pull request by jsuchome**
Extension registry is part of the core service, which is also
installed by the installer. So to access this registry, it's only
necessary to remember where fuseml-core service was installed.

Registering is done by POST calls to the registry, for building
the right payload it is necessary to build a structure that registry
expects. Until it is possible to just import fuseml-core module,
let's copy the necessary structs.

### CODE REVIEW

**Comment by jsuchome** [comment]
As mentioned elsewhere, this does only register with subset of extension description - notably, endpoints and credentials are missing.
For example, this is what is registered after adding `mlflow` extension:
```
> curl -X GET -H "Content-Type: application/x-yaml" http://fuseml-core.172.18.0.2.nip.io/extensions
- id: mlflow
  product: mlflow
  version: 1.19.0
  description: |
    MLFlow is an open source platform specialized in tracking ML experiments, and packaging and deploying ML models.
  status:
    registered: 2021-08-24T13:55:16Z
    updated: 2021-08-24T13:55:16Z
  services:
  - id: mlflow-store
    extension_id: mlflow
    resource: s3
    category: model-store
    auth_required: false
    description: MLFlow minio S3 storage back-end
    status:
      registered: 2021-08-24T13:55:16Z
      updated: 2021-08-24T13:55:16Z
  - id: mlflow-tracking
    extension_id: mlflow
    resource: mlflow-tracking
    category: experiment-tracking
    auth_required: false
    description: MLFlow experiment tracking service API and UI
    status:
      registered: 2021-08-24T13:55:16Z
      updated: 2021-08-24T13:55:16Z
```

File: deployments/extensions.go [hunk 1]
```diff
@@ -466,6 +538,134 @@ func (e *Extension) Uninstall(c *kubernetes.Cluster, ui *ui.UI, options *kuberne
...
+
+// Unregister extension from the extension registry
+func (e *Extension) UnRegister(c *kubernetes.Cluster, ui *ui.UI, options *kubernetes.InstallationOptions) error {
+	domain, err := options.GetString("system_domain", "")
+	if err != nil {
+		return errors.New("system_domain value not provided")
+	}
+
+	fusemlURL := fmt.Sprintf("http://%s.%s", CoreDeploymentID, domain)
+	fullURL := fmt.Sprintf("%s/extensions/%s", fusemlURL, e.Desc.Product)
```
Comment (stefannica) [inline deployments/extensions.go:582]: Same thing here: the extension is identified by its `ID` field, not by its `product` field.

File: deployments/extensions.go [hunk 2]
```diff
@@ -466,6 +538,134 @@ func (e *Extension) Uninstall(c *kubernetes.Cluster, ui *ui.UI, options *kuberne
...
+	if resp.StatusCode == 204 {
+		return nil
+	}
+
+	// no such extension found, that might be OK...
+	if resp.StatusCode == 404 {
+		return nil
+	}
+
+	fmt.Println("response from DELETE: ", resp.StatusCode)
```
Comment (stefannica) [inline deployments/extensions.go:608]: debug leftovers ?

File: deployments/extensions.go [hunk 3]
```diff
@@ -466,6 +538,134 @@ func (e *Extension) Uninstall(c *kubernetes.Cluster, ui *ui.UI, options *kuberne
...
+
+	// something else is wrong, read the response from POST call
```
Comment (stefannica) [inline deployments/extensions.go:610]: I think you meant DELETE call

File: deployments/extensions.go [hunk 4]
```diff
@@ -466,6 +538,134 @@ func (e *Extension) Uninstall(c *kubernetes.Cluster, ui *ui.UI, options *kuberne
...
+	}
+
+	fusemlURL := fmt.Sprintf("http://%s.%s", CoreDeploymentID, domain)
+	fullURL := fmt.Sprintf("%s/extensions", fusemlURL)
+
+	extServices := []*registeredExtensionService{}
+	for _, service := range e.Desc.Services {
+		extSetvice := registeredExtensionService{
+			ID:          service.ID,
+			ExtensionID: &e.Name,
```
Comment (stefannica) [inline deployments/extensions.go:632]: hint: you don't need to set the extension ID for services during registration, the core service can deduce it automatically, because you're registering everything together as a hierarchy of objects

File: deployments/extensions.go [hunk 5]
```diff
@@ -466,6 +538,134 @@ func (e *Extension) Uninstall(c *kubernetes.Cluster, ui *ui.UI, options *kuberne
...
+			Resource:    service.Resource,
+			Category:    service.Category,
+			Description: service.Description,
+		}
```
Comment (stefannica) [inline deployments/extensions.go:636]: Why not do the same for endpoints and even credentials ? The internal endpoints can already be included in the extension descriptor. In some cases, that's all you need.

**Review by stefannica — CHANGES_REQUESTED** [review]

File: deployments/extensions.go [hunk 6]
```diff
@@ -466,6 +538,134 @@ func (e *Extension) Uninstall(c *kubernetes.Cluster, ui *ui.UI, options *kuberne
...
+	if err != nil {
+		return errors.New("system_domain value not provided")
+	}
+
+	fusemlURL := fmt.Sprintf("http://%s.%s", CoreDeploymentID, domain)
+	fullURL := fmt.Sprintf("%s/extensions", fusemlURL)
+
+	extServices := []*registeredExtensionService{}
+	for _, service := range e.Desc.Services {
+		extSetvice := registeredExtensionService{
```
Comment (stefannica) [inline deployments/extensions.go:630]: typo: should be `extService`

File: deployments/extensions.go [hunk 7]
```diff
@@ -466,6 +538,134 @@ func (e *Extension) Uninstall(c *kubernetes.Cluster, ui *ui.UI, options *kuberne
...
+		Version:     &e.Desc.Version,
+		Description: &e.Desc.Description,
+		Services:    extServices,
+	}
+
+	jsonValue, err := json.Marshal(ext)
+	if err != nil {
+		return err
+	}
+	resp, err := http.Post(fullURL, "application/json", bytes.NewBuffer(jsonValue))
```
Comment (stefannica) [inline deployments/extensions.go:652]: This is probably where you wanted to check if the extension is already registered, otherwise you'll get a Conflict response here.
You should consider the case where someone is running the installer repeatedly to install more extensions. Should an "extension already exists" response here be treated as an error ?

File: paas/install_client.go [hunk 8]
```diff
@@ -211,13 +211,24 @@ func (c *InstallClient) handleExtensions(action string, extensions []string, opt
 			if err != nil {
 				return errors.New(fmt.Sprintf("Failed to install extension %s: %s", extension.Name, err.Error()))
 			}
+
+			c.ui.Note().Msg(fmt.Sprintf("Registering extension '%s'...", extension.Name))
+			err = extension.Register(c.kubeClient, c.ui, options)
```
Comment (stefannica) [inline paas/install_client.go:216]: Not all installed extensions need to be registered with FuseML (e.g. cert-manager, KNative).

File: cmd/internal/client/uninstall.go [hunk 9]
```diff
@@ -9,6 +9,13 @@ import (
 )
 
 var uninstallOptions = kubernetes.InstallationOptions{
+	{
+		Name:        "system_domain",
+		Description: "The domain used for FuseML. Should be pointing to the traefik public IP (Leave empty to use a nip.io domain).",
```
[the remark this answers is not in the document]
Comment (jsuchome) [reply cmd/internal/client/uninstall.go:14]: that's just copy and paste of some other text, let me fix it

File: deployments/extensions.go [hunk 10]
```diff
@@ -466,6 +538,134 @@ func (e *Extension) Uninstall(c *kubernetes.Cluster, ui *ui.UI, options *kuberne
 	return nil
 }
 
+// get extension from the registry
+func (e *Extension) getExtension(ui *ui.UI, fusemlURL string) error {
```
[the remark this answers is not in the document]
Comment (jsuchome) [reply deployments/extensions.go:542]: It was there just for testing, but you are right it's not used now. I'll delete it

File: deployments/extensions.go (hunk 6 above)
Comment (jsuchome) [reply deployments/extensions.go:630]: fixed

File: deployments/extensions.go (hunk 2 above)
Comment (jsuchome) [reply deployments/extensions.go:608]: indeed

File: deployments/extensions.go (hunk 1 above)
Comment (jsuchome) [reply deployments/extensions.go:550]: Should be `Name` actually

File: deployments/extensions.go (hunk 4 above)
Comment (jsuchome) [reply deployments/extensions.go:598]: OK

File: paas/install_client.go (hunk 8 above)
Comment (jsuchome) [reply paas/install_client.go:216]: Hm... why not?

Of course they are not needed for fuseml-core operations (they won't be needed from workflows for example), but they are useful to know for the installer, so the user can list what extensions are actually installed and potentially delete some.

After all, we are not calling them ML extensions.

File: deployments/extensions.go (hunk 5 above)
Comment (jsuchome) [reply deployments/extensions.go:602]: What do you mean by "the same"? Adding empty section? 
OK, sure.

File: deployments/extensions.go (hunk 7 above)
Comment (jsuchome) [reply deployments/extensions.go:618]: OK, good point, here's probably the point where my "getExtension" function should be used.

> Should an "extension already exists" response here be treated as an error ?

Hm, what do you think? I think probably not ... for the case of running installer repeatedly it would be a bummer.

File: paas/install_client.go (hunk 8 above)
Comment (jsuchome) [reply paas/install_client.go:216]: Hm... why not?

Of course they are not needed for fuseml-core operations (they won't be needed from workflows for example), but they are useful to know for the installer, so the user can list what extensions are actually installed and potentially delete some.

After all, we are not calling them ML extensions.

File: deployments/extensions.go [hunk 11]
```diff
@@ -466,6 +538,134 @@ func (e *Extension) Uninstall(c *kubernetes.Cluster, ui *ui.UI, options *kuberne
...
+	if resp.StatusCode == 200 {
+
+		var buf bytes.Buffer
+		_, _ = io.Copy(&buf, resp.Body)
+
+		fmt.Printf("response body as string: \n%s\n", buf.String())
+
+		ext := registeredExtension{}
+
+		err = json.Unmarshal(buf.Bytes(), &ext)
```
[the remark this answers is not in the document]
Comment (jsuchome) [reply deployments/extensions.go]: Yep, it needs `yaml` here (but I'm not using it any more for now)

[the remark this answers is not in the document]
Comment (jsuchome) [reply deployments/extensions.go]: Yep, it needs `yaml` here (but I'm not using it any more for now)

File: deployments/extensions.go (hunk 7 above)
Comment (jsuchome) [reply deployments/extensions.go:672]: See my 3rd commit

File: paas/install_client.go (hunk 8 above)
Comment (stefannica) [reply paas/install_client.go:216]: Right now, the only reason you need to register extensions with FuseML is so that you can consume them in workflows. I would say cert-manager has no meaning for ML workflows, and there will be other cases of similar services that are not useful for MLOps. 

Ok, I agree that KNative might be useful for ML workflows, because you could have a workflow component that deploys prediction services directly on top of KNative, so you could advertise/register KNative as a prediction service platform, but that doesn't mean all k8s applications are relevant for ML.

To put it another way, the main purpose of the extension registry is not to provide persistent storage for the installer.

If the fuseml core were also responsible for _installing_ the extensions instead of the installer, like we discussed a few days ago, that would be a completely different story: you would need to store everything in the core.

But if you asked me how can the installer decide which extensions it needs to register with FuseML, I wouldn't be able to give a good answer :)

File: deployments/extensions.go (hunk 7 above)
Comment (stefannica) [reply deployments/extensions.go:672]: Yup, I agree

**Review by stefannica — APPROVED** [review]
Aside from the discussion related to which extensions should be registered with the FuseML core service, everything looks good, thank you.
